### ISSUE TITLE
Feature: Fabric mod metadata parsing

### ISSUE THREAD

**Issue by TehNut** [MultiMC/MultiMC5#2501]
Right now [Fabric](https://fabricmc.net/) mods don't have metadata in the mod view and just show their file name. The metadata file is `fabric.mod.json` in the jar's root dir, just like an `mcmod.info` for Forge. A basic overview of it's spec can be seen [here](https://gist.github.com/TehNut/340f62d4167b225cb6d3f54119fff5a0) and a real-world example can be seen [here](https://github.com/TehNut/HWYLA/blob/eea7c427354255d59b1687b954f2a7813316322e/src/main/resources/fabric.mod.json).

I tried implementing this myself, but I'm not extremely familiar with C++ (nor it's workspace setup/build process), so I figured it would be best to just make a request.

If it helps, here's what I ended up with in my attempt. I don't even know if it compiles 🤷‍♂️ 
<details>
  <summary>Click to expand</summary>

  ```cpp
void Mod::ReadFabricModInfo(QByteArray contents)
{
    QJsonParseError jsonError;
    QJsonDocument jsonDoc = QJsonDocument::fromJson(contents, &jsonError);
    auto obj = jsonDoc.object();
    m_mod_id = obj.value("id").toString();
    m_name = obj.value("name").toString();
    m_version = obj.value("version").toString();
    auto links = obj.value("links");
    m_homeurl = links.isObject() ? links.value("homepage").toString() : links.toString();
    m_description = obj.value("description").toString();

    QJsonArray authors = firstObj.value("authors").toArray();
    if (authors.size() == 0)
        authors = firstObj.value("authors").toArray();

    if (authors.size() == 0)
        m_authors = "";
    else if (authors.size() >= 1)
    {
        m_authors = authors.at(0).toString();
        for (int i = 0; i < authors.size(); i++)
        {
            auto author = authors.at(i);
            if (author.isObject()) {
                m_authors += (m_authors.isEmpty() ? "" : ", ") + author.value("name");
            } else {
                author = author.toString();
                // String format wraps emails in <> and websites in (), so we need to remove those.
                m_authors += (m_authors.isEmpty() ? "" : ", ") + author.remove("\\((.+)\\)").remove("<(.+)>");
            }
        }
    }
}
  ```
</details>

**Comment by asiekierka** [comment]
Fabric mod format support would likely be better off waiting until a stable 1.14 release. Things are still changing, though the fields used in this code sample particular are not likely to change.

**Comment by peterix** [comment]
Also, have you considered a `provides` field?

**Comment by peterix** [comment]
@asiekierka I have some thoughts on this that I'd love to talk about. Mostly it's about getting enough flexibility in such a system to allow fixes to be made afterwards.

`provides` gets you partially there, because someone can go in afterwards and replace a broken dependency with a new one that works. Or, you can have multiple implementations of a dependency.

The ultimate solution is to separate metadata from the data entirely and operate a dependency resolver on a layer separate from what the mods (packages) do.

**Comment by asiekierka** [comment]
>The question would be: is the file authoritative? Does fabric rely on the contents while loading mods? Can I be sure the dependencies listed are not just some made up junk?

Yes, yes and yes. There are also no "hidden" ways of defining the data in the class file or anything of the sort - the JSON is authoritative. (But the specification is not frozen yet)

>Also, have you considered a provides field?

We have some plans of our own in this regard. This mostly revolves around nested "JAR-in-JARs" for providing small dependencies and bundling them without the need for the user to download large amounts of dependencies (especially with Fabric's hopes of promoting increased hook modularity). In this particular scenario, we hope to separate the interface .JAR and the implementation .JAR - as such, mods depend (and bundle, if necessary!) the interface, depend on the interface in a range of specific versions, and implementations bundle the interface always. In addition, the mod loader has agency to pick the best-fitting (usually latest) version of the interface or hook among the bundled mods.

I wonder what you think of this approach to the problem. A different solution (which I just started thinking about during a walk, so apologies if there's a glaring flaw in it) could be enforcing a policy of "one JAR providing a given package", and then using "provides" information or another form of package-based versioning to decide which JAR is routed to for which package.

**Comment by peterix** [comment]
Well, I think of the larger problem - how to apply this sort of thing to already released mods from the past 10 years. Having meaningful metadata inside the mods that has to be correct in order for the mod to work helps, IF it is in place. For the most of Minecraft's modding history, this was not the case. 

For example, I can never implicitly trust metadata inside of Forge mods, because it did not use the metadata for a very long time, and a lot of it is simply wrong. The code in this issue is at the same level -- display only. It leaves a lot on the table.

I'd like to be able to show people errors and resolve dependencies without having to run the game or mod loaders, and I'd like to do it accurately (not like curse, where the dependencies are only on the level of projects). For that to work, the dependency resolving logic I have has to closely match the dependency resolving logic employed by the mod loader(s). It has to be a superset of everything else.

From my POV, I'd prefer to have **very** dumb loaders that can be driven by a separate mod installation/resolver/downloader/launcher component with user friendly UI, because then I don't have to build such a complex superset of all this dependency logic.

In other words, can we have an actually good metadata repository for mods? Ideally something like the KSP CKAN? I don't care about hosting the actual files, but I'd love to be able to provide the metadata next to the mods.

**Comment by asiekierka** [comment]
>Well, I think of the larger problem - how to apply this sort of thing to already released mods from the past 10 years. Having meaningful metadata inside the mods that has to be correct in order for the mod to work helps, IF it is in place. For the most of Minecraft's modding history, this was not the case.

I would recommend https://github.com/asiekierka/modalyzer as a good start here. Unfortunately, Fabric's format as it stands is geared more or less specifically *for* Fabric mods - I don't mind discussing the idea of changing this and making a more general mod metadata format, but I fear such an attempt would just turn into [comic 927](https://xkcd.com/927/).

**Comment by peterix** [comment]
Just having this metadata in the mods and making it mandatory to use is good enough already. I couldn't ask for more. This means I can actually put these mods into a metadata system without a ton of automated checkers.

As for what I think of that XKCD... I'd do it just for the hell of it and if it fails, it fails :D
